### ISSUE TITLE
Provide an option or alternative for --set and/or --set-string to take a value literally

### ISSUE THREAD

**Issue by maxhov** [helm/helm#4030]
I am experiencing an escaping problem with Helm. For instance, if I want to set a password in a chart, I run into the problem that `\` characters are removed.

Example:

1. Take a chart that sets a k8s secret where the value of the secret is base64 encoded by Helm (b64enc)
1. Run `helm upgrade --install --set password=Passw\Ord! somechart`
1. The resulting base64 encoded string decodes to `PasswOrd!` instead of `Passw\Ord!`

I looked into possible solutions on my end and did the following:
1. Singly escaping the `\` resulting in `Passw\\0rd!` does not give the desired output, both `\` are removed, resulting in `PasswOrd!`.
1. Wrapping with either ticks (`'`) or quotes (`"`) results `\` being removed, resulting in `PasswOrd!`.

What did work though is escaping the backslash two times `Passw\\\Ord!`, which results in `Passw\Ord!`.

# Expected behaviour
I would expect that `\\` would result in `\` instead of both being removed. 

# Desired outcome
It should be possible for passwords to contain special characters, such as `\`. 

# Possible solution:
Wrapped variable values in the `--set` argument should be taken literally (e.g. `--set password='Passw\Ord!'` should result in `Passw\Ord!`.

Output of `helm version`:
2.8.2

Output of `kubectl version`:
Client: 1.10.1
Server: 1.9.6

Edit: updated character naming to be more consistent on what is meant

**Comment by mattfarina** [comment]
It appears this is the expected escape behavior. For the code location see https://github.com/kubernetes/helm/blob/b6660cd5c9a9ee35ff24034fcc223b7eaeed3ee2/pkg/strvals/parser.go#L309

For Helm 2 we should document this as changing it would cause a backward incompatibility.

**Comment by bacongobbler** [comment]
labelling this as a good first issue if someone wants to do a deep dive around the --set parser and document their findings.

**Comment by technosophos** [comment]
The escape character for the `--set` parser is `/`.

Char '\\' cited in the code above is the `\` character, escaped per Go's escaping rules.

Keep in mind that you have the shell interpreting things here, too. So often you have to double escape things because the shell will interpret `\` as an escape sequence. Thus, you might need to do `\\` to escape the `\` character for the shell. And in some cases you will need to `\\\\` to escape things. Wrapping the value in single quotes (`--set name='value'`) will also have an impact in most shells. But the bottom line is that you need to consider all the different things that are interpreting the set value between when you execute `helm` and when the final value is injected into Helm's values.

**Comment by muratg** [comment]
I was playing with `stable/registry` chart today, and hit this issue. I was passing `htpasswd` file contents to `--set`. `$` character was the culprit.

Workaround was just escaping it, but finding the root cause took a little while!

**Comment by smurfralf** [comment]
I will work on this as my first contribution to documentation.

**Comment by bacongobbler** [comment]
closed via #4485

**Comment by muratg** [comment]
@smurfralf I believe you'll need to escape `$` character in your example as well as `\`. 

cc @bacongobbler

**Comment by muratg** [comment]
Also, I'm not sure if I agree with the resolution. It's great the documentation is updated (thanks @smurfralf!) but shouldn't the real fix be not just taking quoted values literally? Even with the docs, this will continue to confuse folks.

**Comment by 0x6d6178** [comment]
I would be very much in favor of actually taking quoted values for —set (variants) literally. 
Especially in CI/CD setups escaping (secret) values can be extra challenging, as one might be unaware of the actual value being processed. To overcome this, quite some extra code/parsing of the values would be required in the pipeline before the value can be offered to Helm.

**Comment by muratg** [comment]
I agree with @0x6d6178.

**Comment by moikot** [comment]
I'm not entirely sure that the workaround proposed by @0x6d6178 is going to cover 100% of the cases. How about a password including comma? Do I understand correctly that a comma needs to be escaped using backslash?

**Comment by smurfralf** [comment]
Its not clear to me in the comments above what "quoted values" and "literally" mean. Since this is a CLI we have to contend with the shell no matter what. So chars like dollar sign ($), quote (") and tick(') are going to have to be escaped in some way by the user regardless. The use of ticks to quote for the shell also serves to protect against comma (,) and equals (=) delimiter parsing. So isn't the only thing being debated is whether the yaml interpretation of backslashes should be continue?

**Comment by moikot** [comment]
@smurfralf You absolutely right, the proposed solution (e.g. --set password="Passw\Ord!" should result in Passw\Ord!) is not gonna work since it's not up to Helm to decide how to treat '\O' in double quotes. 

But I'm afraid there is another aspect as well, even if you use a value provided in a variable, you are not safe since a comma (',') and an opening curly brace ('{') has a special meaning in --set and --set-string and you have to escape them. For example:

```bash
export QQQ="bar,baz"
helm lint --set foo=${QQQ}
```

Gives you an error
```bash
Error: failed parsing --set data: key "baz" has no value
```

and in case of '{'

```bash
export QQQ="{bar"
helm lint --set foo=${QQQ}
```

Gives you an error
```bash
Error: failed parsing --set data: list must terminate with '}'
```

There is another issue reported about the escaping and I think it's more correct #4406.

**Comment by 0x6d6178** [comment]
Bash does not interpret strings that are wrapped with ticks ('), whereas it does for some characters when the string is wrapped with quotes (") (see https://stackoverflow.com/questions/6697753/difference-between-single-and-double-quotes-in-bash#6697781). I would expect that a string is not interpreted by Helm when I supply a tick wrapped string in the `--set` option, just like the behaviour of Bash.

Currently the behaviour is:

```bash
helm lint --set foo='{bar'
```
Currently results in:
```bash
Error: failed parsing --set data: list must terminate with '}'
```
Whereas I would expect the value to be taken literally.

However, I understand the additional functionality that `--set` provides (e.g. allowing lists, multiple values etc.), so maybe it would be an idea to introduce a new CLI option (as #4406 also suggests), for instance something like `--set-literal` that does not interpret any of the characters that are supplied by the user. Would that be feasible?

**Comment by anisimovsergey** [comment]
@0x6d6178 Command line processors, in general, remove all the special characters (double quotes (") or single quote (')) before passing the resulting value to a program, so Helm doesn't get a chance to distinguish between them.

**Comment by elchtestAtBosch** [comment]
I just ran into the same problem with passwords containing a comma and I agree with 0x6d6178 that I would expect an option to take the string literally. I tried set-string and it would make sense for me if that would take the string literally but it doesn't. I am also in the situation where the problem occurs in the CI pipeline and adding escaping logic to the pipeline seems unnecessary.

**Comment by elchtestAtBosch** [comment]
I would suggest a better title: "Provide an option for --set to take value literally"

**Comment by technosophos** [comment]
@0x6d6178 @elchtestAtBosch  the reason you are getting that parse error is because of this:

https://github.com/helm/helm/blob/master/docs/using_helm.md#the-format-and-limitations-of---set

When you pass a value with `--set`, it goes through a type conversion routine to determine what type the value is. The `{}` syntax is the list syntax, and is legal in a value (so is a comma). Keep in mind that when you single-quote a string in your shell (at least in most shells), it appears to the invoked program in exactly the same format as an unquoted value: `--set foo=bar` and `--set foo='bar'` both come into the executable as `foo = "bar"`. So there is no way for a program to say " if it is single quoted in the shell, that means...", because the quotes are stripped off.

So if you don't want the value to be interpolated, use `--set-string` and it will be treated as a string. Or you can use backslash escapes.

**Comment by moikot** [comment]
With all my due respect, @technosophos, but `--set-string` exhibits the same issue as `--set`

```bash
helm lint --set-string foo="baz,bar"
Error: failed parsing --set-string data: key "bar" has no value
```

as well as 

```
helm lint --set-string foo="{baz"
Error: failed parsing --set-string data: list must terminate with '}'
```

The only option to get a comma or an open brace (a the first character) is to use escaping.

**Comment by elchtestAtBosch** [comment]
To use --set-string would be fine and intuitive. Unfortunately I tried it and it didn't work. Throws the same error.

**Comment by bacongobbler** [comment]
There are two ways to avoid both bash escaping and the --set parser's syntax with curly braces and commas. You can use `--values/-f` by supplying a values file with those parameters, or by using `--set-file foo=bar.txt`. Anything other than those two CLI flags will require bash escape codes or escape codes for the `--set` parser. We can implement a `--set-literal` but then you're still going to have to fight with your terminal to escape double quotes/single quotes and the like, which kinda defeats the purpose.

To use `--set-file`:

```
$ echo "bar,baz" > bar.txt
$ helm install [...] --set-file foo=bar.txt
```

And for `--values`:

echo "foo: bar" > values.yaml
$ helm install [...] -f values.yaml
```

Hope that helps.

**Comment by elchtestAtBosch** [comment]
Thx, @bacongobbler for the tips. We worked around it by banning special characters from our passwords.

Bash is not the problem it has the "take it literally" feature. But Helm is not the only tool that stumbles upon special characters in passwords so we decided to avoid this entirely.

I would still love to see --set-string behave as at least I would expect to take the value as a string and not parse it. Afaik the purpose of set-string is to avoid integer parsing. This kind of falls into the same category.

**Comment by bacongobbler** [comment]
It doesn't fall into the same category because it was intentionally designed to work this way from the beginning.

When --set-string was first introduced, it was built with the intention of coercing values into their string types, without sacrificing the existing --set parser syntax so they can still allocate arrays on the command line, e.g. `--set-string list={1,2,3}`. It's only intention was to work around a limitation in --set where a number would be coerced into scientific notation due to the YAML parser.

Because of the existing flags I laid out that work around the issue, I don't believe we plan on supporting a --set-literal flag in Helm 2. In Helm 3, we can take a closer look at the --set parser overall and see what we can do to improve (especially with regards to the comma syntax), so the feedback provided here is useful. Thanks for providing your feedback.

**Comment by 0x6d6178** [comment]
> I would suggest a better title: "Provide an option for --set to take value literally"

Good suggestion. Given the discussion's direction I changed the title to reflect the discussion more clearly.

**Comment by AuroreM** [comment]
Hi @maxhov and @bacongobbler !
Is a contribution about adding a --litteral option is still welcomed ? I would be glad to work on this as a first PR :)

**Comment by hickeyma** [comment]
@AuroreM I'd say go for it. We can see then when reviewing the PR.

**Comment by tredmon** [comment]
Want to add that this just bit me. I needed to set a string value obtained via REST call that held a JSON object. `{"key": "value"}` it instead ended up as `["key": "value"]` which of course failed during my service's attempt to parse as JSON.

I tried to fix using the `--set-file "myKey=<(printf %s "$JSON_OBJECT")"` using bash to generate a file descriptor, but helm failed to read the file. It seems the only workaround is to actually write to a file. 😢 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x2188d79]

goroutine 1 [running]:
main.readFile(0xc0003d10e0, 0x1d, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0xc00044a100, 0xc0006834e0, ...)
```

**Comment by bacongobbler** [comment]
Did you try escaping the brackets via `\{"key": "value"\}`? You could also try `--set-string` if that doesn't work for you.

https://helm.sh/docs/intro/using_helm/#the-format-and-limitations-of-set notes that `{1,2,3}` is the format for setting array values, so that seems in line with the documentation.

**Comment by tredmon** [comment]
I run helm in an automated fashion (CI/CD), and as I said the JSON_OBJECT is the result of a REST call. I suppose I could escape the brackets and commas using `sed`, but that could get complex if string values within the JSON contain brackets or commas.

**Comment by j-coll** [comment]
Any updates on this?

**Comment by technosophos** [comment]
Update on what? If you want to PR a change, I'm sure it would be welcomed. To my knowledge, nobody is actively working on it. But most likely, you just need to use escape sequences as described above. (Note that in many of these cases, it is _the shell_ that is interpreting characters, not Helm)

**Comment by cm0s** [comment]
I know `--set password=Passw\Ord! `is just an example, but I'm not sure it's really safe to try to pass a password with the `helm install` command. It will be easy to get access to it with a `helm get values [RELEASE_NAME]`..

**Comment by PSanetra** [comment]
Is there any documentation which characters have to be escaped for passing a string literally to `--set` or `--set-string`? I need to pass a JSON string as a string value to helm.

**Comment by bacongobbler** [comment]
https://helm.sh/docs/intro/using_helm/#the-format-and-limitations-of---set

**Comment by PSanetra** [comment]
@bacongobbler On that page it is only mentioned that you need to escape "special" characters "sometimes", but there is no list of characters, which are considered special.

**Comment by brianharwell** [comment]
The ability to pass a literal string would be very nice!

**Comment by muratg** [comment]
@brianharwell I came to change my mind on this issue. I believe there's nothing to be done but document it properly. I agree with @PSanetra that an exact list of characters to be escaped would be nice to document.

There's no such "literal string" per se, as you're in the shell and certain characters are expected to be escaped when used in arguments. Try writing a toy cmdline tool (perhaps with node.js) and see if you can parse "literal" strings.

**Comment by bacongobbler** [comment]
@tredmon started listing out most of the special characters earlier in this thread: https://github.com/helm/helm/issues/4030#issuecomment-562263937

I know of a few more to add to that list.

AFAIK, the list of special characters used in the set parser most people are interested in are the following:

```
=
{
[
,
.
]
}
```

Depending on whether you care about typed values or not, the literal strings `null`, `true`, `false`, and numerical numbers are treated as typed values, and caster either as integers, booleans, or null values. Then there's [floating values](https://github.com/helm/helm/issues/1707).

Asides that, I'm not aware of any others, but if someone wants to double-check, the set parser lives in [pkg/strvals](https://github.com/helm/helm/blob/master/pkg/strvals/parser.go).

More info:

https://github.com/helm/helm/blob/b6bbe4f08bbb98eadd6c9cd726b08a5c639908b3/pkg/strvals/parser.go#L159
https://github.com/helm/helm/blob/b6bbe4f08bbb98eadd6c9cd726b08a5c639908b3/pkg/strvals/parser.go#L263
https://github.com/helm/helm/blob/b6bbe4f08bbb98eadd6c9cd726b08a5c639908b3/pkg/strvals/parser.go#L276
https://github.com/helm/helm/blob/b6bbe4f08bbb98eadd6c9cd726b08a5c639908b3/pkg/strvals/parser.go#L339
https://github.com/helm/helm/blob/b6bbe4f08bbb98eadd6c9cd726b08a5c639908b3/pkg/strvals/parser.go#L356

We'd appreciate PRs if someone wants to write that up in the document I linked to earlier. There are a few examples shown above in earlier comments.

Hope this helps.

**Comment by brianharwell** [comment]
> @brianharwell I came to change my mind on this issue. I believe there's nothing to be done but document it properly. I agree with @PSanetra that an exact list of characters to be escaped would be nice to document.
> 
> There's no such "literal string" per se, as you're in the shell and certain characters are expected to be escaped when used in arguments. Try writing a toy cmdline tool (perhaps with node.js) and see if you can parse "literal" strings.

This application is written in Go, yes? The Docker Cli and the Kubernetes Cli (kubectl) are both written in Go and they both support string literals. How do they do it?

I have actually never encountered a command line application that did not support wrapping arguments in double quotations or needed me to escape special characters. I'm primarily an MS developer and when I wrap args with the `"` they are escaped properly. 🤷

**Comment by muratg** [comment]
@brianharwell Are you able to use "$" for example, inside quotes?

I'm not on Windows currently, so I undertand it may be different there, but in linux/macos shell "$" means "expand the variable on the right of it".  So you need to escape it if you'd like to use the character "$".

Small example:

```sh
~ x="lol"  
~ echo "haha $x" 
haha lol
~ echo "haha \$x"
haha $x
```

**Comment by bacongobbler** [comment]
Powershell/cmd.exe uses the dollar sign syntax as well to refer to variables, so it would also have to be escaped, yes.

**Comment by brianharwell** [comment]
In Windows, it works differently. To use a set variable you would do `%TEMP%` so...

`echo "Path: %TEMP%`

and with Powershell, it would be...

`Write-Host "Path ${env:Temp}"`

In your example the `$x` is being expanded but if you wrap the echo in single quotes like so...

`echo 'haha $x'`

It does not escape the value.

It makes sense the shell is doing the expansion, but helm shouldn't. If I try passing a value to `--set` using one or all of the special tokens helm should just take the value provided by the shell.

So in your example...

```
~ x="lol"
~ echo "haha $x"
haha lol
~ echo 'haha $x'
haha $x
~ helm --set lol="haha $x"
sets lol to "haha lol"
 ~ helm --set lol='haha $x'
sets lol to "haha $x"

**Comment by brianharwell** [comment]
@bacongobbler 

Yes, it does so in powershell...

```
~ $x="lol"
~ echo "haha $x"
haha lol
~ echo 'haha $x'
haha $x
```

**Comment by muratg** [comment]
@brianharwell Which character are you having issues with currently? 

I only had issues with `$` and `\` before I moved all those settings to config files.

If helm indeed does further parsing on the string that shell has passed, I'd concur that it's a problem. But that can obviously be fixed/amended.

**Comment by bacongobbler** [comment]
https://github.com/helm/helm/issues/4030#issuecomment-427568512

**Comment by brianharwell** [comment]
@bacongobbler Thanks for linking that comment. 

https://github.com/helm/helm/issues/4030#issuecomment-640234355

All the more reason to add `--set-literal` so that `--set-literal foo='{bar'` allows bypassing any additional parsing that helm does to parameters.

Essentially, what I am advocating for is the ability to pass `{$foo,].bar` and to use that value in my templates with no escaping necessary. The implementation details I will leave for the fine developers of helm to determine.

**Comment by bacongobbler** [comment]
From another earlier comment: https://github.com/helm/helm/issues/4030#issuecomment-624356805

> If you want to PR a change, I'm sure it would be welcomed. To my knowledge, nobody is actively working on it.

We've been pretty clear here: we will accept pull requests from the community wiling to put in the time to introduce this feature flag. Nobody's arguing that this should not be a Helm feature. Someone just has to put in the time to implement it.

**Comment by brianharwell** [comment]
I understand and I would help out if the app was .Net Core. :-) Unfortunately, I know zero Go.

**Comment by andrewmcguinness** [comment]
I'm afraid I don't understand why it wouldn't be correct for  `--set-string` to have this behaviour.

The documentation says `--set-string` "sets a string value"

When I do

```
--set-string variable="{one}two"
```

it does not set a string value. It sets an array value

`helm get values` outputs

```
variable:
  - one
```
instead of

```
variable: '{one}two'
```

Finally note that if I wanted the array, I can (now) do the less ambiguous

```
--set string variable[0]=one
```

**Comment by bacongobbler** [comment]
`--set-string` was introduced to handle cases with the YAML parser, where you wanted the value to be handled as a string, like `--set-string foo.myports={8080, 8081, 8082)`. It still handles the input the same as `--set` such that you can define arrays and dictionaries similar to `--set`.

It is uncommon, but also we've found cases where the first 7 characters of a git commit ID are all numerical, and these are often desired to be coerced as a string value rather than an integer.

Consider that `1234567e10` is also treated as a common floating number (`e10` is scientific notation), and `NO` is treated as a boolean value.

In other words, `--set-string` was introduced to handle a separate set of cases than what's being proposed here. Your argument is not wrong, but the use case it was originally designed to handle (retain `--set`-like syntax but force the YAML parser to coerce the value as a string) differs than what you're proposing. Changing the behaviour of `--set-string` now would break those users currently relying on that functionality.

Hope this helps.

**Comment by brianharwell** [comment]
Any update on adding `--set-literal` or something similar?

**Comment by pscheid92** [comment]
Hello all together :wave: I would like to give it a try.\
I just opened a draft PR for visibility for experiments (no working implementation, yet).

I added a `--set-literal` argument and duplicated the existing parser code; stripped away some functions/methods and added a small test case. I first have to familiarize myself with the existing parser code.

Currently, escaping is coupled heavily with the parser in `parser.go`, therefore a duplicated parser in `literal_parser.go`.\
Maybe we can merge both eventually, we will see.

**Comment by dennisotugo** [comment]
This is still an issue

**Comment by PSanetra** [comment]
@bacongobbler in the comment https://github.com/helm/helm/issues/4030#issuecomment-644444107 you did include this list as the set of characters which should be escaped when passing a string to set-string:
> [...]
> 
> ```
> =
> {
> [
> ,
> .
> ]
> }
> ```
> [...]

In this list the character `\` is missing, which should be escaped, too. Can you edit your comment and also include `\`?

**Comment by brianonn** [comment]
I ran into this issue today trying to set a literal json string to use it in a template (to set an env var).  We get the json string from our TUF repository root key and I just want to set it as a json literal string in an environment variable in the k8s container spec.
`--set ` or `--set-string` both try to parse the string.  I can really use a `--set-literal key='{"foo":"bar"}'`

**Comment by Aitozi** [comment]
+1 for the this feature :)

**Comment by gm42** [comment]
> https://helm.sh/docs/intro/using_helm/#the-format-and-limitations-of---set

Since this page still does **not** list the characters that must be escaped, nor does the CLI, shall I make a PR to address the issue?

**Comment by dennisotugo** [comment]
+1

**Comment by Amirilw** [comment]
+1

**Comment by manoj-bandara** [comment]
+1 for a feature like --set-literal

this issue of not taking literals in --set is causing many problems for us

**Comment by bkalcho** [comment]
+1 for `--set-literal`

This will be really useful. When you need to pass stringified JSON through GitHub actions to helm it is hell, as you need to escape the GH YAML templating engine, bash, and at the end Helm interpretation.

**Comment by natalicot** [comment]
+1 for a feature like --set-literal

**Comment by amkartashov** [comment]
I expect `--set-string` to actually do its job. Why on Earth there is a need for escaping? It's a string, there is no need to parse it in a first place.

**Comment by paymog** [comment]
+1 to `set-literal` and also requesting a `set-literal-sensitive` so that passwords and such can also be set literally without worrying about leaking them.
